File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.py
def solution(phone_book):
    answer = True
    phone_book.sort(key=len)
    dicA = {}
    dicB = {}
    lstA = []
    lstB = []
    length = []
    for i in range(len(phone_book)):
        if len(phone_book[i]) in length:
            continue
        else:
            for j in range(i, len(phone_book)):
                if len(phone_book[j]) == len(phone_book[i]):
                    lstA.append(phone_book[j][:len(phone_book[i])])
                else:
                    lstB.append(phone_book[j][:len(phone_book[i])])
            dicA[len(phone_book[i])] = lstA
            dicB[len(phone_book[i])] = lstB
            lstA = []
            lstB = []
            length.append(len(phone_book[i]))

    for k in range(len(length)):
        tempA = dicA[length[k]]
        tempB = dicB[length[k]]
        tempB = list(set(tempB))
        temp_total = tempA + tempB
        if len(temp_total) != len(set(temp_total)):
            answer = False
            break

    return answer

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.java
import java.util.*;

class Solution {
    public static boolean solution(String[] phone_book) {
        boolean answer = true;
        Arrays.sort(phone_book, Comparator.comparingInt(String::length));
        ArrayList<Integer> length = new ArrayList<>();
        HashMap<Integer, ArrayList<String>> dicA = new HashMap<>();
        HashMap<Integer, ArrayList<String>> dicB = new HashMap<>();

        for (int i = 0; i < phone_book.length; i++) {
            int len = phone_book[i].length();
            if (length.contains(len)) continue;

            ArrayList<String> lstA = new ArrayList<>();
            ArrayList<String> lstB = new ArrayList<>();

            for (int j = i; j < phone_book.length; j++) {
                if (phone_book[j].length() == len) {
                    lstA.add(phone_book[j].substring(0, len));
                } else {
                    lstB.add(phone_book[j].substring(0, len));
                }
            }

            dicA.put(len, lstA);
            dicB.put(len, lstB);
            length.add(len);
        }

        for (int len : length) {
            ArrayList<String> tempA = dicA.get(len);
            ArrayList<String> tempB = dicB.get(len);
            HashSet<String> tempBSet = new HashSet<>(tempB);
            ArrayList<String> tempTotal = new ArrayList<>(tempA);
            tempTotal.addAll(tempBSet);

            if (tempTotal.size() != new HashSet<>(tempTotal).size()) {
                answer = false;
                break;
            }
        }

        return answer;
    }
}
